File: python/sparklabs/batch/sales/rdd/product_rank.py
from sparklabs.batch.sales.rdd import utils


def run() -> None:
    rows = (
        utils.order_items.map(lambda item: (item["product_id"], 1))
        .reduceByKey(lambda left, right: left + right)
        .map(lambda item: (item[0], 1))
        .join(utils.products.map(lambda product: (product["product_id"], product["category"])))
        .map(lambda item: (item[1][1], 1))
        .reduceByKey(lambda left, right: left + right)
        .sortBy(lambda item: item[1], ascending=False)
        .collect()
    )
    for index, row in enumerate(rows, start=1):
        print((row[0], row[1], index))


if __name__ == "__main__":
    run()

File: scala/src/main/scala/io/devorbit/sparklabs/batch/sales/rdd/ProductRank.scala
package io.devorbit.sparklabs.batch.sales.rdd

object ProductRank extends App {
  val categoryCounts = RddUtils.orderItems
    .map(item => (item.product_id, 1))
    .reduceByKey(_ + _)
    .map { case (productId, _) => (productId, 1) }
    .join(RddUtils.products.map(product => (product.product_id, product.category)))
    .map { case (_, (_, category)) => (category, 1) }
    .reduceByKey(_ + _)
    .sortBy(_._2, ascending = false)
    .collect()

  categoryCounts.zipWithIndex.foreach { case ((category, count), index) =>
    println((category, count, index + 1))
  }
}
